core/packages/contracts/src/Gateway.sol

```diff
@@ -2,35 +2,476 @@
+            catch {
+                success = false;
+            }
+        } else if (message.command == COMMAND_UPGRADE) {
+            try Gateway(this).upgrade{gas: GAS_FOR_DISPATCH}(message.params) {}
+            catch {
+                success = false;
+            }
+        }
+
+        emit IGateway.InboundMessageDispatched(message.origin, message.nonce, success);
+    }
+
+    /**
+     * Getters
+     */
+
+    function operatingMode() external view returns (OperatingMode) {
+        return CoreStorage.layout().mode;
+    }
+
+    function channelOperatingModeOf(ParaID paraID) external view returns (OperatingMode) {
+        CoreStorage.Layout storage $ = CoreStorage.layout();
+        return $.channels[paraID].mode;
+    }
 
-    constructor(Registry registry) RegistryLookup(registry) {
-        _setRoleAdmin(SENDER_ROLE, ADMIN_ROLE);
+    function channelNoncesOf(ParaID paraID) external view returns (uint64, uint64) {
+        Channel storage ch = CoreStorage.layout().channels[paraID];
+        return (ch.inboundNonce, ch.outboundNonce);
     }
 
-    function handle(ParaID origin, bytes calldata message) external virtual;
+    function channelFeeRewardOf(ParaID paraID) external view returns (uint256, uint256) {
+        Channel storage ch = CoreStorage.layout().channels[paraID];
+        return (ch.fee, ch.reward);
+    }
+
+    function agentOf(bytes32 agentID) external view returns (address) {
+        CoreStorage.Layout storage $ = CoreStorage.layout();
+        return $.agents[agentID];
+    }
+
+    /**
+     * Handlers
+     */
+
+    struct AgentExecuteParams {
+        bytes32 agentID;
+        bytes payload;
+    }
+
+    // Execute code within an agent
+    function agentExecute(bytes calldata data) external onlySelf {
+        CoreStorage.Layout storage $ = CoreStorage.layout();
+
+        AgentExecuteParams memory params = abi.decode(data, (AgentExecuteParams));
 
-    function outboundQueue() internal view returns (IOutboundQueue) {
-        return IOutboundQueue(resolve(OUTBOUND_QUEUE));
+        address agent = $.agents[params.agentID];
+        if (agent == address(0)) {
+            revert AgentDoesNotExist();
+        }
+
+        if (params.payload.length == 0) {
+            revert InvalidAgentExecutionPayload();
+        }
+
+        bytes memory call = abi.encodeCall(AgentExecutor.execute, (address(this), params.payload));
+        try Agent(payable(agent)).invoke($.agentExecutor, call) {}
+        catch {
+            revert AgentExecutionFailed();
+        }
+    }
+
+    struct CreateAgentParams {
+        bytes32 agentID;
     }
 
-    function ensureOrigin(ParaID a, ParaID b) internal pure {
-        if (a != b) {
+    // Create an agent for a consensus system on Polkadot
+    function createAgent(bytes calldata data) external onlySelf {
+        CoreStorage.Layout storage $ = CoreStorage.layout();
+
+        CreateAgentParams memory params = abi.decode(data, (CreateAgentParams));
```

Comment:
Can we remove the separate decode step?
```suggestion
    function createAgent(CreateAgentParams calldata params) external onlySelf {
        CoreStorage.Layout storage $ = CoreStorage.layout();
```

Reply:
`createAgent` is actually called by `submitInbound`. Since `abi.decode` can error and revert the entire transaction, it's safer to decode within `createAgent`.

Notice that `submitInbound` calls `createAgent` within a try catch block. However in Solidity, try catch is not able to handle errors from abi.decode.